### ISSUE TITLE
[Feature request] Easier way to specify parameters to freeze

### ISSUE THREAD

**Issue by mfkasim1** [patrick-kidger/equinox#31]
Thank you for this library! This is the library that makes most sense among JAX's nn libraries (I don't have to learn additional transformations on those libs).
What I found quite tedious at the moment is specifying parameters to freeze.
In your example, the filter has to be specified based on the structure of the model.
This becomes very tedious if we want to try out different architectures as we have to write the filter for each model.
In PyTorch, we can use buffer in this case. In equinox, there is no easy way to have something like buffer.

**Option 1**
One option is to use `eqx.static_field`, but this removes the field from PyTree which is not what we always want.

**Option 2**
Another option is to use `jax.lax.stop_gradient`, this works as below:

```python
class Module(eqx.Module):
    mult: jnp.ndarray
    mult_b: jnp.ndarray  # let's say this is the buffer

    def __init__(self, mult: jnp.ndarray, mult_b: jnp.ndarray) -> None:
        self.mult = mult
        self.mult_b = mult_b  # using jax.lax.stop_gradient(mult_b) doesn't work here

    def __call__(self, x: jnp.ndarray) -> jnp.ndarray:
        x = self.mult * jax.lax.stop_gradient(self.mult_b) * x
        return x

if __name__ == "__main__":
    key = jax.random.PRNGKey(123)
    key1, key2, keyx = jax.random.split(key, 3)
    mult = jax.random.normal(key1, (1,))
    mult_b = jax.random.normal(key2, (1,))
    x = jax.random.normal(keyx, (1,))
    module = Module(mult, mult_b)
    out = module(x)

    grad = eqx.filter_grad(lambda mod, x: jnp.sum(module.__class__.__call__(mod, x)))
    g = grad(module, x)
    print(g)  # we want the gradient of mult_b is 0 or None
```

The example above works, but it is quite cumbersome to write `jax.lax.stop_gradient` in every method in the class.

**Option 3**
Another option is to use filter by field name (maybe something like `eqx.filter_by_field_name`). I imagine that goes something like this:

```python
class Module(eqx.Module):
    mult: jnp.ndarray
    mult_b: jnp.ndarray  # let's say this is the buffer

    def __init__(self, mult: jnp.ndarray, mult_b: jnp.ndarray) -> None:
        self.mult = mult
        self.mult_b = mult_b

    def __call__(self, x: jnp.ndarray) -> jnp.ndarray:
        x = self.mult * self.mult_b * x
        return x

if __name__ == "__main__":
    key = jax.random.PRNGKey(123)
    key1, key2, keyx = jax.random.split(key, 3)
    mult = jax.random.normal(key1, (1,))
    mult_b = jax.random.normal(key2, (1,))
    x = jax.random.normal(keyx, (1,))
    module = Module(mult, mult_b)
    out = module(x)

    name_regex = r"\w+_b"
    filter_spec = jax.tree_map(lambda _: False, (module, x))
    filter_spec = eqx.filter_by_field_name(filter_spec, name_regex, replace=True)
    grad = eqx.filter_grad(lambda mod, x: jnp.sum(module.__class__.__call__(mod, x)),
                           filter_spec=filter_spec)
    g = grad(module, x)
    print(g)  # we want the gradient of mult_b is 0 or None
```

I would vote for the 3rd option, because it is the most convenient, but it would need an additional function: `eqx.filter_by_field_name`
What do you think?
I'm happy to create a PR if you agree with this.

**Comment by patrick-kidger** [comment]
So you've raised what makes for an interesting point about how Equinox/JAX works!

Could we specify parameters by field name? The answer is no -- and here's why.

(And don't worry, I'll provide a couple of solutions afterwards. ;) )

The key fact is that **Equinox modules are not privileged**. Your proposal would mean that `eqx.filter_by_field_name` parses its input PyTree, and has special handling whenever it encounters an `eqx.Module`. (It looks for field names and then sets things appropriately.)

However the central feature of Equinox is the fact that an `eqx.Module` is _never_ privileged in any way. There is no coupling between `eqx.Module` and any other code in Equinox. No function in Equinox ever does an `isinstance(..., eqx.Module)` check. Overall, `eqx.Module` is not in any way more special than a classical list-of-dicts or whatever.

This is the key fact that makes Equinox code easy to reason about, especially when passing it through lots of JAX transformations.

---

Okay, what are the solutions? Here are two possibilities.

**Possibility 1: Make the attribute a `@property`:**

```python
class Module(eqx.Module):
    _mult_b: jnp.ndarray

    @property
    def mult_b(self):
        return jax.lax.stop_gradient(self._mult_b)

    def __call__(self, x: jnp.ndarray) -> jnp.ndarray:
        return self.mult_b * x  # neat syntax, hurrah
```
Just put the `lax.stop_gradient` in one place. Then they don't need to be peppered throughout the whole code, nor do you need to be concerned that you might have forgotten one somewhere.

This is essentially a neatened version of your "Option 2".

(Entirely in passing: note that I haven't provided an `__init__` method here. If you're just directly setting attributes then Equinox will autogenerate an `__init__` method that does that for you. i.e. `Module(jnp.array(...))` will work here.)

**Possibility 2: use `is_leaf=...`**

```python
# Module with buffers
class MyModule(eqx.Module):
    buffer: jnp.ndarray
    ...

# Maybe MyModule is used as part of some larger module
class SomeModule(eqx.Module):
    modules: List[MyModule]
    ...

module = SomeModule(...)
is_mymodule = lambda x: isinstance(x, MyModule)
get_mymodule = lambda m: [x for x in jax.tree_leaves(m, is_leaf=is_mymodule) if is_mymodule(x)]
get_buffers = lambda m: [x.buffer for x in get_mymodule(m)]
filter_spec = jax.tree_map(lambda _: True, module)
filter_spec = eqx.tree_at(get_buffers, module, replace=False)
```

This time, we've got something that is basically equivalent to your "Option 3", it just uses existing functionality to explicitly locate all buffers anywhere in the PyTree. Note that `jax.tree_leaves(..., is_leaf=...)` is relatively new functionality in JAX. If you're on an older version of JAX then you'll need to write `jax.tree_flatten(..., is_leaf=...)[0]` instead.

---

Do either of those solve the problem for you?

For what it's worth, I actually have an even-better-still solution prepared. But it needs a (minor) bug to be fixed in core JAX first, namely https://github.com/google/jax/issues/9538.

Once that's sorted then

```python
class Buffer(eqx.Module):
    array: jnp.ndarray

    def __jax_array__(self):
        return lax.stop_gradient(array)

class Module(eqx.Module):
    buffer: jnp.ndarray

    def __init__(self, buffer):
        self.buffer = Buffer(buffer)
```

should work seamlessly.

**Comment by mfkasim1** [comment]
Thanks for your comprehensive answer. Currently the option 2 works for me, but I'm looking forward for the last solution to get the bug fixed!

**Comment by danielward27** [comment]
Hi Patrick,

Sorry for bringing up an old thread.

- Possibility 1 is great when you know ahead of time that you don't want to train the parameters, but more challenging in other cases, e.g. when you don't write  the class yourself.

- Possibility 2 is more flexible, but in my case there may be non-trainable, and trainable modules of the same type.
 
I could make a wrapper class ``NonTrainableModule``, to aid filtering/to apply stop gradient, but that also feels clunky - it would be a pain to write wrappers separately for each abstract module, and have to forward all the methods.

In an ideal world, I'd want to an Array wrapper similar to ``Buffer`` above, acting exactly like an array but 1. having a different type that I can use in filtering, and ideally also 2. calling ``stop_gradient``. This would allow me to mark the submodule (or specific parameters of a submodule) as non-trainable directly after instantiation (e.g. using ``eqx.tree_at``, or ``jax.tree_util.tree_map``), which would be much nicer imo.  However I see that ``__jax_array__`` is likely to be removed. 

Do you have any recommendations?

**Comment by patrick-kidger** [comment]
Indeed, I no longer recommend using `__jax_array__` at all. Fortunately, it turns out we don't actually need it. Here's a quick demo:
```python
# Library code

class Wrapper(eqx.Module):
    array: jax.Array

def _is_wrapper(x):
    return isinstance(x, Wrapper)

def _unwrap(x):
    if _is_wrapper(x):
        return jax.lax.stop_gradient(x.array)
    else:
        return x

def unwrap(x: PyTree):
    return jax.tree_util.tree_map(_unwrap, x, is_leaf=_is_wrapper)

# User code

model = ...
model = eqx.tree_at(lambda m: m.path.to.field, model, replace_fn=Wrapper)

@jax.grad
def do_stuff(model, ...):
    model = unwrap(model)
    ...

do_stuff(model, ...)
```
that is, just insert your custom markers anywhere you like, then process them however you like. [The same trick can be used for custom parameterisations, like symmetric matrices](https://docs.kidger.site/equinox/tricks/#custom-per-parameter-behaviour).

**Comment by danielward27** [comment]
Amazing. Thanks for the help. I modified it a bit in case I need to unwrap something which contains another unwrappable. 

```python
from abc import abstractmethod
from typing import Any

import equinox as eqx
import jax
import jax.numpy as jnp
from jax import Array, lax

PyTree = Any


class AbstractUnwrappable(eqx.Module):

    def recursive_unwrap(self):
        """Returns the unwrapped pytree, unwrapping subnodes as required."""
        flat, tree_def = eqx.tree_flatten_one_level(self)
        tree = jax.tree.unflatten(tree_def, recursive_unwrap(flat))
        return tree.unwrap()

    @abstractmethod
    def unwrap(self):
        """Returns the unwrapped pytree, assuming no subnodes need to be unwrapped."""
        pass


class Symmetric(AbstractUnwrappable):
    matrix: Array

    def unwrap(self):
        return 0.5 * (self.matrix + self.matrix.T)


class NonTrainable(AbstractUnwrappable):
    tree: PyTree

    def unwrap(self):
        differentiable, static = eqx.partition(self.tree, eqx.is_array_like)
        return eqx.combine(lax.stop_gradient(differentiable), static)


def recursive_unwrap(x: PyTree):
    return jax.tree_util.tree_map(
        f=lambda leaf: (
            leaf.recursive_unwrap() if isinstance(leaf, AbstractUnwrappable) else leaf
        ),
        tree=x,
        is_leaf=lambda x: isinstance(x, AbstractUnwrappable),
    )


class MyModule(eqx.Module):
    arr: Array


module = MyModule(
    Symmetric(NonTrainable(jnp.arange(9, dtype=float).reshape(3, 3))),
)


@jax.grad
def loss(module):
    module = recursive_unwrap(module)
    return jnp.sum(module.arr)


assert jnp.all(recursive_unwrap(loss(module)).arr == jnp.zeros((3, 3)))
```
